core/src/cluster_info_repair_listener.rs

```diff
@@ -0,0 +1,683 @@
+                let _ = Self::recv_loop(
+                    &blocktree,
+                    &mut peer_roots,
+                    exit,
+                    &cluster_info,
+                    epoch_schedule,
+                );
+            })
+            .unwrap();
+        Self {
+            thread_hdls: vec![thread],
+        }
+    }
+
+    fn recv_loop(
+        blocktree: &Blocktree,
+        peer_roots: &mut HashMap<Pubkey, (u64, u64)>,
+        exit: Arc<AtomicBool>,
+        cluster_info: &Arc<RwLock<ClusterInfo>>,
+        epoch_schedule: EpochSchedule,
+    ) -> Result<()> {
+        let socket = UdpSocket::bind("0.0.0.0:0").unwrap();
+        let my_id = cluster_info.read().unwrap().id();
+        let mut my_gossiped_root = 0;
+
+        loop {
+            if exit.load(Ordering::Relaxed) {
+                return Ok(());
+            }
+
+            let peers = cluster_info.read().unwrap().gossip_peers();
+            let mut peers_needing_repairs: HashMap<Pubkey, EpochSlots> = HashMap::new();
+
+            // Iterate through all the known nodes in the network, looking for ones that
+            // need repairs
+            for peer in peers {
+                let last_update_ts = Self::get_last_ts(peer.id, peer_roots);
+                let my_root =
+                    Self::update_my_gossiped_root(&my_id, cluster_info, &mut my_gossiped_root);
+                {
+                    let r_cluster_info = cluster_info.read().unwrap();
+
+                    // Update our local map with the updated peers' information
+                    if let Some((peer_epoch_slots, ts)) =
+                        r_cluster_info.get_epoch_state_for_node(&peer.id, last_update_ts)
+                    {
+                        // Following logic needs to be fast because it holds the lock
+                        // preventing updates on gossip
+                        peer_roots.insert(peer.id, (ts, peer_epoch_slots.root));
+                        if Self::should_repair_peer(
+                            my_root,
+                            peer_epoch_slots.root,
+                            &epoch_schedule,
+                            NUM_BUFFER_SLOTS,
+                        ) {
+                            // Clone out EpochSlots structure to avoid holding lock on gossip
+                            peers_needing_repairs.insert(peer.id, peer_epoch_slots.clone());
+                        }
+                    }
+                }
+            }
+
+            // After updating all the peers, send out repairs to those that need it
+            let _ = Self::serve_repairs(
+                &my_id,
+                blocktree,
+                peer_roots,
+                &peers_needing_repairs,
+                &socket,
+                cluster_info,
+                &epoch_schedule,
+                &mut my_gossiped_root,
+            );
+
+            sleep(Duration::from_millis(REPAIRMEN_SLEEP_MILLIS as u64));
```

Comment:
is this the model we use for gossiped votes, too?

Reply:
yeah